src/kiota.core/Writers/JavaWriter.cs

```diff
@@ -85,20 +86,55 @@ public override void WriteIndexer(CodeIndexer code)
         public override void WriteMethod(CodeMethod code)
         {
             //TODO javadoc
-            WriteLine("@javax.annotation.Nonnull");
-            WriteLine($"public {(code.IsAsync ? "java.util.concurrent.Future<" : string.Empty)}{GetTypeString(code.ReturnType).ToFirstCharacterUpperCase()}{(code.IsAsync ? ">" : string.Empty)} {code.Name.ToFirstCharacterLowerCase()}({string.Join(',', code.Parameters.Select(p=> GetParameterSignature(p)).ToList())}) {{ return null; }}");
+            WriteLine(code.ReturnType.IsNullable ? "@javax.annotation.Nullable" : "@javax.annotation.Nonnull");
+            WriteLine($"{GetAccessModifier(code.Access)} {(code.IsAsync ? "java.util.concurrent.CompletableFuture<" : string.Empty)}{GetTypeString(code.ReturnType).ToFirstCharacterUpperCase()}{(code.IsAsync ? ">" : string.Empty)} {code.Name.ToFirstCharacterLowerCase()}({string.Join(',', code.Parameters.Select(p=> GetParameterSignature(p)).ToList())}) {{");
+            IncreaseIndent();
+            switch(code.MethodKind) {
+                case CodeMethodKind.IndexerBackwardCompatibility:
+                    WriteLine($"final {code.ReturnType.Name} builder = new {code.ReturnType.Name}();");
+                    WriteLine($"builder.currentPath = this.currentPath + this.{pathSegmentPropertyName} + \"/\" + position;");
+                    WriteLine("return builder;");
+                break;
+                default:
+                    WriteLine("return null;");
+                break;
+            }
+            DecreaseIndent();
+            WriteLine("}");
         }
-
+        private const string pathSegmentPropertyName = "pathSegment";
+        private const string currentPathPropertyName = "currentPath";
         public override void WriteProperty(CodeProperty code)
         {
             //TODO: missing javadoc
-            WriteLine("@javax.annotation.Nullable");
-            WriteLine($"public {GetTypeString(code.Type)} {code.Name};");
+            var returnType = GetTypeString(code.Type);
```

Comment:
I'm assuming at some point we need to break these out into separate methods.

Reply:
Yeah I'm doing that in the next pull requests